tests/mail/tests.py

```diff
@@ -40,48 +48,105 @@
     HAS_AIOSMTPD = False
 
 
-class HeadersCheckMixin:
+def message_from_bytes(s):
+    """email.message_from_bytes() using modern email.policy.default."""
+    return _message_from_bytes(s, policy=policy.default)
+
+
+def message_from_binary_file(fp):
+    """email.message_from_binary_file() using modern email.policy.default."""
+    return _message_from_binary_file(fp, policy=policy.default)
+
+
+class MailTestsMixin:
     def assertMessageHasHeaders(self, message, headers):
         """
         Asserts that the `message` has all `headers`.
 
-        message: can be an instance of an email.Message subclass or a string
+        message: can be an instance of an email.Message subclass or bytes
                  with the contents of an email message.
         headers: should be a set of (header-name, header-value) tuples.
         """
         if isinstance(message, bytes):
             message = message_from_bytes(message)
         msg_headers = set(message.items())
-        self.assertTrue(
-            headers.issubset(msg_headers),
-            msg="Message is missing "
-            "the following headers: %s" % (headers - msg_headers),
-        )
+        if not headers.issubset(msg_headers):
+            missing = "\n".join(f"  {h}: {v}" for h, v in headers - msg_headers)
+            actual = "\n".join(f"  {h}: {v}" for h, v in msg_headers)
+            raise self.failureException(
+                f"Expected headers not found in message.\n"
+                f"Missing headers:\n{missing}\n"
+                f"Actual headers:\n{actual}"
+            )
 
+    def assertStartsWith(self, first, second):
+        if not first.startswith(second):
+            self.longMessage = True
+            self.assertEqual(
+                first[: len(second)],
+                second,
+                "First string doesn't start with the second.",
+            )
```

Comment:
Do we need custom assertions for such checks? Is it not enough to use
```python
assertTrue(first.startswith(second), "First string doesn't start with the second.")
```
or even:
```python
assertTrue(first.startswith(second))
```

Reply:
assertStartsWith() [already existed later](https://github.com/django/django/blob/519087819ed6e8bfbe6be208df71a7df19f23a58/tests/mail/tests.py#L1336-L1343) in the file; I just hoisted it to the mixin. (And added a companion assertEndsWith().)

The custom assertion vs assertTrue() makes no difference when the tests are passing. But when a test fails, the custom assertion outputs context that's important for understanding why. (This is especially helpful when, say, a test only fails on a platform that uses \r\n lineseps, but the developer is working on a \n platform and can't easily poke around in a debugger on the platform where the test is failing. Just, um, for example. 😄 )